examples/src/main/java/io/grpc/examples/manualflowcontrol/ManualFlowControlServer.java

```diff
@@ -0,0 +1,147 @@
+
+                                            String message = "Hello " + name;
+                                            System.out.println("<-- " + message);
+                                            HelloReply reply = HelloReply.newBuilder().setMessage(message).build();
+
+                                            // Send a response.
+                                            responseObserver.onNext(reply);
+                                        } else {
+                                            System.out.println("Done.");
+                                            responseObserver.onCompleted();
+                                        }
+                                    }
+                                }
+                            }
+                        });
+                    }
+                });
+
+                // Give gRPC a StreamObserver it can write incoming requests into.
+                return new StreamObserver<HelloRequest>() {
+                    @Override
+                    public void onNext(HelloRequest request) {
+                        // Process the request and send a response or an error.
+                        try {
+                            // Accept and enqueue the request.
+                            String name = request.getName();
+                            System.out.println("--> " + name);
+                            work.add(name);
+                            // Signal the sender to send another request.
+                            serverCallStreamObserver.request(1);
+                        } catch (Throwable throwable) {
+                            throwable.printStackTrace();
+                            responseObserver.onError(Status.UNKNOWN.withCause(throwable).asException());
+                        }
+                    }
+
+                    @Override
+                    public void onError(Throwable t) {
+                        // End the response stream if the client presents an error.
+                        t.printStackTrace();
+                        responseObserver.onCompleted();
+                    }
+
+                    @Override
+                    public void onCompleted() {
+                        // Signal the end of work when the client ends the request stream.
+                        work.add(COMPLETED);
+                    }
+                };
+            }
+        };
+
+        final Server server = ServerBuilder
+                .forPort(50051)
+                .addService(svc)
+                .build()
+                .start();
+
+        System.out.println("Listening on " + server.getPort());
+
+        Runtime.getRuntime().addShutdownHook(new Thread() {
+            @Override
+            public void run() {
+                System.out.println("Shutting down");
+                server.shutdown();
+                pool.shutdown();
```

Comment:
technically, pool needs to await server termination since it could add more events to the pool after server shutdown